Vagrantfile

```diff
@@ -0,0 +1,161 @@
+# -*- mode: ruby -*-
+# vi: set ft=ruby :
+
+# https://github.com/hashicorp/vagrant/issues/9442#issuecomment-374785457
+unless Vagrant::DEFAULT_SERVER_URL.frozen?
+  Vagrant::DEFAULT_SERVER_URL.replace('https://vagrantcloud.com')
+end
+
+Vagrant.configure("2") do |config|
+  # VM Box
+  config.vm.box = "ubuntu/bionic64"
+  # Automatic box update checking
+  config.vm.box_check_update = true
+
+  # CodeIgniter virtual host
+  config.vm.network "forwarded_port", guest: 80, host: 8080, host_ip: "127.0.0.1"
+  # Code Coverage virtual host
+  config.vm.network "forwarded_port", guest: 81, host: 8081, host_ip: "127.0.0.1"
+  # User Guide virtual host
+  config.vm.network "forwarded_port", guest: 82, host: 8082, host_ip: "127.0.0.1"
+  # MySQL server
+  #config.vm.network "forwarded_port", guest: 3306, host: 3307, host_ip: "127.0.0.1"
+  # PostgreSQL server
+  #config.vm.network "forwarded_port", guest: 5432, host: 5433, host_ip: "127.0.0.1"
+  # Memcached server
+  #config.vm.network "forwarded_port", guest: 11211, host: 11212, host_ip: "127.0.0.1"
+  # Redis server
+  #config.vm.network "forwarded_port", guest: 6379, host: 6380, host_ip: "127.0.0.1"
+
+  # Add "192.168.10.10 ${VIRTUALHOST}" in your host file to access by domain
+  #config.vm.network "private_network", ip: "192.168.10.10"
+
+  # Same path set in the $CODEIGNITER_PATH Provision
+  # "virtualbox" type allow auto-sync host to guest and guest to host
+  # but chmod does not work... tests will fail.
+  # Default rsync__args except "--copy-links", to allow phpunit correctly works by symlink
+  config.vm.synced_folder ".", "/var/www/codeigniter", type: "rsync", rsync__args: ["--verbose", "--archive", "--delete", "-z"]
+
+  # Provider-specific configuration
+  config.vm.provider "virtualbox" do |vb|
+    # Display the VirtualBox GUI when booting the machine
+    vb.gui = false
+    # Customize the amount of memory on the VM:
+    vb.memory = "768"
```

Comment:
VM runs with 400MB - 308MB "stopped". Required ~600MB when running tests

Reply:
I run mine with 1024m. Probably overkill, but hoping/planning for near-complete code coverage eventually.
Is there much of a penalty specifying more memory?

Reply:
It depends on the user's machine. Thinking about taking up as little resources as possible, I started with 512MB, but PHPUnit broke for not having enough memory to do the coverage. I run again and saw that it occupies about 600MB at most (MySQLi driver). I increased 256MB. 

Testing now with the SQLite3 driver, PHPUnit also could not gererate the code coverage:

```
Type:        ErrorException
Message:     proc_open(): fork failed - Cannot allocate memory
Filename:    /var/www/codeigniter/vendor/sebastian/version/src/Version.php
Line Number: 89
```

I'll go up to 1024 and test again. But this seems to me a very high value, I'm thinking of maybe already letting in the script instructions to create a swap file, because the memory only rises so much when it's generating code coverage.

Reply:
SQLite3 with `database.tests.database = ':memory:'`...

Reply:
OK. Increased the default `vb.memory` to 1024 and added script to auto add 1GB of swap if the VM has low RAM.

Tested using SQLite3 with `vb.memory = "512"` and no problem. It used ~120MB of swap during coverage.